pkg/plugins/manager.go

```diff
@@ -0,0 +1,170 @@
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-plugin"
+	"github.com/treeverse/lakefs/pkg/logging"
+)
+
+var (
+	errPluginGroupNotFound = fmt.Errorf("unknown plugin group")
+	errPluginTypeNotFound  = fmt.Errorf("unknown plugin type")
+)
+
+// PluginType specifies the type of related plugins
+// For example, "diff" is a type of plugins that know how to perform a diff.
+type PluginType string
+
+// PluginName specifies the name of a plugin
+// For example, "delta" is a plugin.
+type PluginName string
+
+// PluginIdentity identifies the plugin implementation.
+type PluginIdentity struct {
+	Impl               plugin.Plugin
+	ImplName           string
+	HandshakeConfig    plugin.HandshakeConfig
+	ExecutableLocation string
+}
+type pluginTypeConfigMap map[PluginName]plugin.ClientConfig
+
+// PluginTypeNameMap example:
+/*
+{
+	"diff": {
+		"delta": plugin.clientConfig{
+			HandshakeConfig: plugin.HandshakeConfig{
+				ProtocolVersion:  1,
+				MagicCookieKey:   "key",
+				MagicCookieValue: "value",
+			},
+			Cmd: exec.Command("path/to/executable/diff/delta"),
+			AllowedProtocols: []plugin.Protocol{
+				plugin.ProtocolGRPC,
+			},
+			Impl: map[string]plugin.Plugin{
+				"deltaDiff": DeltaGRPCPlugin,
+			},
+		},
+		"iceberg: {
+			HandshakeConfig: plugin.HandshakeConfig{
+				ProtocolVersion:  1,
+				MagicCookieKey:   "key",
+				MagicCookieValue: "value",
+			},
+			Cmd: exec.Command("path/to/executable/diff/iceberg"),
+			AllowedProtocols: []plugin.Protocol{
+				plugin.ProtocolGRPC,
+			},
+			Impl: map[string]plugin.Plugin{
+				"icebergDiff": IcebergGRPCPlugin
+			},
+		}
+	},
+	"merge": {
+		"delta": {
+			HandshakeConfig: plugin.HandshakeConfig{
+				ProtocolVersion:  1,
+				MagicCookieKey:   "key",
+				MagicCookieValue: "value",
+			},
+			Cmd: exec.Command("path/to/executable/merge/delta"),
+			AllowedProtocols: []plugin.Protocol{
+				plugin.ProtocolGRPC,
+			},
+			Impl: map[string]plugin.Plugin{
+				"deltaMerge": DeltaGRPCPlugin,
+			},
+		},
+		"iceberg: {
+			HandshakeConfig: plugin.HandshakeConfig{
+				ProtocolVersion:  1,
+				MagicCookieKey:   "key",
+				MagicCookieValue: "value",
+			},
+			Cmd: exec.Command("path/to/executable/merge/iceberg"),
+			AllowedProtocols: []plugin.Protocol{
+				plugin.ProtocolGRPC,
+			},
+			Impl: map[string]plugin.Plugin{
+				"icebergMerge": IcebergGRPCPlugin
+			},
+		}
+	},
+}
+*/
+type PluginTypeNameMap map[PluginType]pluginTypeConfigMap
+
+// The Manager holds the different types of plugins that can be used in the plugin system
+type Manager struct {
+	pluginTypes PluginTypeNameMap
+}
+
+// WrapPlugin generates a Wrapper that wraps the go-plugin client.
+//
+// It accepts two parameters: the top PluginType and the required PluginName under it.
+// For example, PluginType = "diff", PluginName = "delta" will generate a Wrapper with a client that performs
+// diffs over Delta Lake tables.
+// It also returns a DestroyClientFunc
```

Comment:
Can we rename Wrapper?  Similarly, not sure where DestroyClientFunc is returned.